### PR TITLE
Nodejs letschat

### PR DESCRIPTION
**Pull request by octaviansoldea**
Dear Reviewer, Hello Hajime and Octavian (Purdila)

Please find enclosed changes I propose for enabling Node.js application Let's Chat to work with LKL and DPDK. Could you please tell if you can consider them for a future release?

I am looking forward to hearing from you.

Thank you very much in advance,
  Octavian (Soldea)

### CODE REVIEW

File: tools/lkl/lib/hijack/hijack.c [hunk 1]
```diff
@@ -291,7 +304,25 @@ int select(int nfds, fd_set *r, fd_set *w, fd_set *e, struct timeval *t)
 	return lkl_call(__lkl__NR_select, 5, nfds, r, w, e, t);
 }
 
-HOOK_CALL_USE_HOST_BEFORE_START(epoll_create);
+HOST_CALL(epoll_create);
+int epoll_create(int flags)
```
Comment (thehajime) [inline tools/lkl/lib/hijack/hijack.c:308]: could you describe (justify) why the previous `HOOK_CALL_USE_HOST_BEFORE_START` isn't enough and why `HOST_CALL` is required ?

File: tools/lkl/lib/virtio_net_dpdk.c [hunk 2]
```diff
@@ -33,9 +33,13 @@ static char *ealargs[4] = {
 /* XXX: disable cache due to no thread-safe on mempool cache. */
 #define MEMPOOL_CACHE_SZ        0
 /* for TSO pkt */
-#define MAX_PACKET_SZ           (65535 \
+/*#define MAX_PACKET_SZ           (65535 \*/
+#define MAX_PACKET_SZ           (16383 \
 	- (sizeof(struct rte_mbuf) + RTE_PKTMBUF_HEADROOM))
+#if 0
```
Comment (thehajime) [inline tools/lkl/lib/virtio_net_dpdk.c]: remove `#if 0`

File: tools/lkl/lib/virtio_net_dpdk.c [hunk 3]
```diff
@@ -33,9 +33,13 @@ static char *ealargs[4] = {
...
 #define MBUF_NUM                (512*2) /* vmxnet3 requires 1024 */
+#endif //0
+#define MBUF_NUM                4096
```
Comment (thehajime) [inline tools/lkl/lib/virtio_net_dpdk.c]: which NIC (driver) does this `MBUF_NUM` value work ?
maybe those are device and hugepage pool size dependent.

File: tools/lkl/scripts/dpdk-sdk-build.sh [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
 dpdk_version="17.02"
 
-git clone -b v${dpdk_version} git://dpdk.org/dpdk dpdk-${dpdk_version}
+#git clone -b v${dpdk_version} git://dpdk.org/dpdk dpdk-${dpdk_version}
```
Comment (thehajime) [inline tools/lkl/scripts/dpdk-sdk-build.sh]: ah, I see this point but you can work around with `${HOME}/.gitconfig` by insteadof statement (https://git-scm.com/docs/git-config#git-config-urlltbasegtinsteadOf). 

So I would prefer to remove this.

**Review by thehajime — CHANGES_REQUESTED** [review]
Thanks for the PR.

I think this commit can be separated into multiple: 1) dpdk related commit, and 2) node.js related one.

plus, an example script to execute letschat application would be helpful.

**Comment by octaviansoldea** [comment]
Hi Hajime

Thank you for your message. This is just an acknowledgment I have received your message. I will come back with an answer asap.

Best regards and wishing you a nice week-end!
     Octavian (Soldea)

File: tools/lkl/scripts/dpdk-sdk-build.sh (hunk 4 above)
Comment (octaviansoldea) [inline tools/lkl/scripts/dpdk-sdk-build.sh]: Hajime, thank you for this comment. Your solution with git is very nice.

File: tools/lkl/lib/virtio_net_dpdk.c (hunk 3 above)
Comment (octaviansoldea) [inline tools/lkl/lib/virtio_net_dpdk.c]: Hajime, the configuration of NICs I use is
 
00:19.0 Ethernet controller: Intel Corporation 82579V Gigabit Network Connection (rev 06)
02:00.0 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
02:00.1 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
02:00.2 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
02:00.3 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
03:00.0 Ethernet controller: Intel Corporation 82574L Gigabit Network Connection
 
where 00:19.0 is connected to LAN (to the wall .... ). In this context, 02:00.0, 02:00.1, 02:00.2, and 02:00.3 are 4 LANs that relies on a same physical NIC, and 03:00.0 is a separate NIC.

In this context, I have connected an inverted internet cable between 02:00.0 and 03:00.0. Does this answer your question?

File: tools/lkl/scripts/dpdk-sdk-build.sh (hunk 4 above)
Comment (octaviansoldea) [reply tools/lkl/scripts/dpdk-sdk-build.sh]: Thank you for this indication. Your solution with git is very nice. I totally agree with this.

File: tools/lkl/lib/virtio_net_dpdk.c (hunk 3 above)
Comment (octaviansoldea) [reply tools/lkl/lib/virtio_net_dpdk.c]: The configuration of NICs I use is

00:19.0 Ethernet controller: Intel Corporation 82579V Gigabit Network Connection (rev 06)
02:00.0 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
02:00.1 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
02:00.2 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
02:00.3 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
03:00.0 Ethernet controller: Intel Corporation 82574L Gigabit Network Connection
where 00:19.0 is connected to LAN (to the wall .... ). In this context, 02:00.0, 02:00.1, 02:00.2, and 02:00.3 are 4 LANs that relies on a same physical NIC, and 03:00.0 is a separate NIC.

In this context, I have connected an inverted internet cable between 02:00.0 and 03:00.0. Does this answer your question?

The truth is that I wanted to keep the default constants and insert ones that worked for me. Do you have any suggestions?

File: tools/lkl/lib/virtio_net_dpdk.c (hunk 2 above)
Comment (octaviansoldea) [reply tools/lkl/lib/virtio_net_dpdk.c]: Please see next question and answer.

**Comment by octaviansoldea** [comment]
Regarding the main comment, I am writing a blog. The blog is supposed to include a walk-through of all the steps. And this will include also code related to node.js. The intention is to make it publicly available as soon as is in a reasonable shape. In any case, I will provide any documentation I can.

Regarding the two parts, I do not fully understand. What do you mean dpdk related commit? The lkl layer does not work without the hook related modifications. Then, the constants are related to connecting lkl dpdk. Do you mean, you would like the changes to be submitted in two different commits?

File: tools/lkl/lib/hijack/hijack.c (hunk 1 above)
Comment (octaviansoldea) [reply tools/lkl/lib/hijack/hijack.c:308]: I will come back on this asap.

File: tools/lkl/lib/virtio_net_dpdk.c (hunk 3 above)
Comment (thehajime) [reply tools/lkl/lib/virtio_net_dpdk.c]: thanks, so you're using Intel i350 NIC (igb driver).
The previous value (512*2) works for my testing (ixgbe and vmxnet3 drivers).

I am not aware of a transparent solution of this variable for the moment.  In such case, comment lines stating is would be nice.

I will also look at existing examples (of DPDK) how they handle this.

**Comment by thehajime** [comment]
> Regarding the main comment, I am writing a blog. The blog is supposed to include a walk-through of all the steps. And this will include also code related to node.js. The intention is to make it publicly available as soon as is in a reasonable shape. In any case, I will provide any documentation I can.

A blog is perfect !
We could put the link from our wiki (https://github.com/lkl/linux/wiki/).

> Regarding the two parts, I do not fully understand. What do you mean dpdk related commit? The lkl layer does not work without the hook related modifications. Then, the constants are related to connecting lkl dpdk. Do you mean, you would like the changes to be submitted in two different commits?

Yes I mean two different commits.

More specifically, changes to `hijack.c` would be a single commit, which improves a support of node.js. And then another commit includes changes to `virtio_net_dpdk.c` which works well with your environment (maybe Intel i350 == igb driver dependent).

I believe those two changes are independent so, I suggested to split your patch.

**Comment by thehajime** [comment]
once you have updated your patches based on the comment, please `git push --force` your patch on the same branch: this pull request should reflect those changes automatically.

File: tools/lkl/lib/hijack/hijack.c (hunk 1 above)
Comment (octaviansoldea) [reply tools/lkl/lib/hijack/hijack.c:308]: Hi Hajime, Dear Reviewers,

The previous version, i.e. with HOOK_CALL_USE_HOST_BEFORE_START, works when using lkl pure. However, it is necessary to use HOST_CALL for the dpdk version. With dpdk and HOOK_CALL_USE_HOST_BEFORE_START I get the message

/home/octavian/Programs/Nodes/node_system_calls/node/install/usr/local/bin/node: symbol lookup error: /home/octavian/Octavian/DriveNetwork/2017/07July/28/lkl-linux/tools/lkl/bin/../liblkl-hijack.so: undefined symbol: host_epoll_create

This means most probably that epoll_create is necessary on the host version.

Please let me know if this answers your question.

Best regards,
  Octavian (Soldea)

Comment (thehajime) [reply tools/lkl/lib/hijack/hijack.c:308]: thanks for the explanation.

`HOOK_CALL_USE_HOST_BEFORE_START` was needed before when we used `epoll_create` in virtio_net_fd.c.  Nowadays we don't use and it should be safe now.

**Comment by thehajime** [comment]
now this branch is upstreamed (https://github.com/lkl/linux/pull/361).

I think this PR and #8 should be submitted to lkl/linux: it is much easier to be integrated.

File: tools/lkl/lib/hijack/hijack.c (hunk 1 above)
Comment (octaviansoldea) [reply tools/lkl/lib/hijack/hijack.c:308]: Hi Hajime

I submitted a new pull request with HOST_CALL to the https://github.com/lkl/linux. My assumption is that DPDK needs epoll_create. Could you please help or provide a bit more feedback?

Thank you in advance,
  Octavian S

Comment (octaviansoldea) [reply tools/lkl/lib/hijack/hijack.c:308]: Hi Hajime

You were perfectly right. In the end it worked. Sorry for my misunderstanding.
I submitted the changes to https://github.com/lkl/linux.

Best regards,
  Octavian Soldea

**Pull request closed without merging**